File: vrpc/core/presence.py
from __future__ import annotations

from ..constants import FALLBACK_LARGE_IMAGE, GITHUB_URL
from ..i18n import t
from ..riot.content import Content
from .state import GameState


def _level_text(state: GameState, language: str) -> str:
    return f"{t(language, 'level')} {state.account_level}"


def build(state: GameState, settings, content: Content, language: str, start_ts: int) -> dict | None:
    if state.session_state == "idle":
        return None

    if state.session_state == "ingame":
        presence = _ingame(state, settings, content, language)
    elif state.session_state == "pregame":
        presence = _pregame(state, settings, content, language)
    else:
        presence = _menu(state, settings, content, language)

    presence["large_text"] = "by yefeblgn"

    if settings.show_elapsed:
        presence["start"] = start_ts

    # "Lobiye Katıl" daveti — Discord '+' menüsünden kanala gönderilebilir.
    # join secret + party (içinde yer olacak şekilde: current < max) gerekir.
    # Discord secret + buttons'ı AYNI ANDA kabul etmediği için davet aktifken
    # GitHub butonu gönderilmez (ikisi birlikte tüm RPC'yi kırar).
    if state.card_id:
        cur = state.party_size if state.party_size > 0 else 1
        mx = state.party_max if state.party_max > cur else cur + 1
        presence["party_id"] = f"vrpc_{state.card_id}"
        presence["party_size"] = [cur, mx]
        presence["join"] = f"vrpc_join_{state.card_id}"
    else:
        if settings.show_party and state.party_size > 0:
            presence["party_size"] = [state.party_size, state.party_max]
        presence["buttons"] = [{"label": t(language, "github"), "url": GITHUB_URL}]

    return presence


def _large_card(state: GameState, content: Content, language: str) -> tuple[str, str]:
    text = _level_text(state, language) if state.account_level else t(language, "app_title")
    return FALLBACK_LARGE_IMAGE, text


def _rank_small(state: GameState, settings, content: Content) -> tuple[str | None, str]:
    icon = content.tier_icon(state.competitive_tier)
    name = content.tier_name(state.competitive_tier)
    rr = state.rr
    if rr is not None and name:
        return icon, f"{name} · {rr} RR"
    return icon, name


def _menu(state: GameState, settings, content: Content, language: str) -> dict:
    large_image, large_text = _large_card(state, content, language)
    details = content.mode_name(state.queue_id) if state.queue_id else t(language, "p_in_menu")

    presence = {"details": details, "large_image": large_image, "large_text": large_text}

    if state.is_queuing:
        presence["state"] = t(language, "queuing")

    _RANK_MODES = {"competitive", "unrated", "premier"}
    mode_key = content._queue_key(state.queue_id)

    if settings.show_rank and state.competitive_tier > 0:
        icon, text = _rank_small(state, settings, content)
        if icon:
            presence["small_image"] = icon
            presence["small_text"] = text
    elif mode_key in _RANK_MODES:
        # Kompakt rank rozeti: tier>0 → rank ikonu, tier==0 → Derecesiz rozeti
        tier = state.competitive_tier if state.competitive_tier > 0 else 0
        icon = content.tier_icon(tier)
        if icon:
            presence["small_image"] = icon
            presence["small_text"] = content.mode_name(state.queue_id)
    elif state.card_id:
        card_sq = content.card_square(state.card_id)
        if card_sq:
            presence["small_image"] = card_sq
            presence["small_text"] = f"{state.name}#{state.tag}" if state.name else ""
    else:
        icon = content.mode_icon_unique(state.queue_id)
        if icon:
            presence["small_image"] = icon
            presence["small_text"] = content.mode_name(state.queue_id)
    return presence


def _pregame(state: GameState, settings, content: Content, language: str) -> dict:
    map_name, splash = content.map_info(state.map_path)
    large_image, large_text = (splash, map_name) if splash else _large_card(state, content, language)

    presence = {"details": t(language, "p_selecting_agent"), "large_image": large_image, "large_text": large_text}

    agent_icon = content.agent_icon(state.agent_uuid)
    agent_name = content.agent_name(state.agent_uuid)
    if agent_icon:
        presence["small_image"] = agent_icon
        presence["small_text"] = agent_name or ""
        if agent_name:
            presence["state"] = agent_name
    else:
        icon = content.tier_icon(state.competitive_tier or 0)
        if icon:
            presence["small_image"] = icon
            presence["small_text"] = t(language, "p_pregame_small")
    return presence


def _ingame(state: GameState, settings, content: Content, language: str) -> dict:
    map_name, splash = content.map_info(state.map_path)
    large_image, large_text = (splash, map_name) if splash else _large_card(state, content, language)

    details = content.mode_name(state.queue_id)

    if state.ally_score is not None and state.enemy_score is not None:
        state_line = f"{t(language, 'p_in_game')} · {state.ally_score} - {state.enemy_score}"
    else:
        state_line = t(language, "p_in_game")

    presence = {
        "details": details,
        "state":   state_line,
        "large_image": large_image,
        "large_text":  large_text,
    }

    agent_icon = content.agent_icon(state.agent_uuid)
    agent_name = content.agent_name(state.agent_uuid)
    if agent_icon:
        presence["small_image"] = agent_icon
        presence["small_text"]  = agent_name or content.mode_name(state.queue_id)
    else:
        icon = content.mode_icon_unique(state.queue_id)
        if icon:
            presence["small_image"] = icon
            presence["small_text"]  = content.mode_name(state.queue_id)
    return presence

File: electron/core/presence.ts
import type {
  BuiltPresence,
  DisplayInfo,
  GameState,
  Language,
  LargeTextMode,
  Settings,
  SmallImageMode,
  SmallTextMode
} from '@shared/types'
import { FALLBACK_LARGE_IMAGE } from '../constants'
import { t } from '@shared/i18n'
import type { Content } from '../riot/content'
import { isCustom, isQueuing } from './state'

function levelText(state: GameState, language: Language): string {
  return `${t(language, 'level')} ${state.accountLevel}`
}

function largeCard(
  state: GameState,
  settings: Settings,
  language: Language
): [string, string] {
  const text =
    settings.showLevel && state.accountLevel
      ? levelText(state, language)
      : t(language, 'app_title')
  return [FALLBACK_LARGE_IMAGE, text]
}

function rankSmall(state: GameState, content: Content): [string | null, string] {
  const icon = content.tierIcon(state.competitiveTier)
  const name = content.tierName(state.competitiveTier)
  if (state.rr !== null && name) return [icon, `${name} · ${state.rr} RR`]
  return [icon, name]
}


function buildMenu(
  state: GameState,
  settings: Settings,
  content: Content,
  language: Language
): BuiltPresence {
  let largeImage: string
  let largeText: string
  if (state.cardId) {
    largeImage = content.cardSquare(state.cardId) || FALLBACK_LARGE_IMAGE
    largeText = state.name ? `${state.name}#${state.tag}` : t(language, 'app_title')
  } else {
    ;[largeImage, largeText] = largeCard(state, settings, language)
  }

  const details = isCustom(state)
    ? t(language, 'custom')
    : state.queueId
      ? content.modeName(state.queueId)
      : t(language, 'p_in_menu')

  const p: BuiltPresence = { details, largeImage, largeText }
  if (isQueuing(state)) p.state = t(language, 'queuing')

  const isComp = content.queueKey(state.queueId) === 'competitive'
  if (isComp) {
    if (settings.showRank && state.competitiveTier > 0) {
      const [icon, text] = rankSmall(state, content)
      if (icon) {
        p.smallImage = icon
        p.smallText = text
      }
    } else {
      const icon = content.tierIcon(state.competitiveTier > 0 ? state.competitiveTier : 0)
      if (icon) {
        p.smallImage = icon
        p.smallText = content.modeName(state.queueId)
      }
    }
  } else {
    const icon = content.modeIcon(state.queueId)
    if (icon) {
      p.smallImage = icon
      p.smallText = content.modeName(state.queueId)
    }
  }
  return p
}

function buildPregame(
  state: GameState,
  settings: Settings,
  content: Content,
  language: Language
): BuiltPresence {
  const [mapName, splash] = content.mapInfo(state.mapPath)
  const [largeImage, largeText] = splash
    ? [splash, mapName]
    : largeCard(state, settings, language)

  const p: BuiltPresence = {
    details: t(language, 'p_selecting_agent'),
    largeImage,
    largeText
  }

  const agentIcon = content.agentIcon(state.agentUuid)
  const agentName = content.agentName(state.agentUuid)
  if (agentIcon) {
    p.smallImage = agentIcon
    p.smallText = agentName || ''
    if (agentName) p.state = agentName
  } else {
    const icon = content.tierIcon(0)
    if (icon) {
      p.smallImage = icon
      p.smallText = t(language, 'p_selecting_agent')
    }
  }
  return p
}

function buildIngame(
  state: GameState,
  settings: Settings,
  content: Content,
  language: Language
): BuiltPresence {
  const [mapName, splash] = content.mapInfo(state.mapPath)
  const [largeImage, largeText] = splash
    ? [splash, mapName]
    : largeCard(state, settings, language)

  const details = isCustom(state) ? t(language, 'custom') : content.modeName(state.queueId)

  const showScore =
    settings.showScore && state.allyScore !== null && state.enemyScore !== null
  const stateLine = showScore
    ? `${t(language, 'p_in_game')} · ${state.allyScore} - ${state.enemyScore}`
    : t(language, 'p_in_game')

  const p: BuiltPresence = { details, state: stateLine, largeImage, largeText }

  const agentIcon = content.agentIcon(state.agentUuid)
  const agentName = content.agentName(state.agentUuid)
  if (agentIcon) {
    p.smallImage = agentIcon
    p.smallText = agentName || content.modeName(state.queueId)
  } else {
    const icon = content.modeIconUnique(state.queueId)
    if (icon) {
      p.smallImage = icon
      p.smallText = content.modeName(state.queueId)
    }
  }
  return p
}


function imageFor(
  mode: Exclude<SmallImageMode, 'auto' | 'none'>,
  state: GameState,
  content: Content
): string | null {
  switch (mode) {
    case 'map':
      return content.mapInfo(state.mapPath)[1]
    case 'agent':
      return content.agentIcon(state.agentUuid)
    case 'rank':
      return content.tierIcon(state.competitiveTier > 0 ? state.competitiveTier : 0)
    case 'playercard':
      return content.cardSquare(state.cardId)
    case 'mode':
      return content.modeIcon(state.queueId)
    default:
      return null
  }
}

function textFor(
  mode: Exclude<LargeTextMode | SmallTextMode, 'auto'>,
  state: GameState,
  content: Content,
  language: Language
): string {
  switch (mode) {
    case 'playerName':
      return state.name ? `${state.name}#${state.tag}` : ''
    case 'level':
      return state.accountLevel ? levelText(state, language) : ''
    case 'mode':
      return content.modeName(state.queueId)
    case 'mapName':
      return content.mapInfo(state.mapPath)[0]
    case 'agentName':
      return content.agentName(state.agentUuid)
    case 'score':
      return state.allyScore !== null && state.enemyScore !== null
        ? `${state.allyScore} - ${state.enemyScore}`
        : ''
    case 'rank': {
      const name = content.tierName(state.competitiveTier)
      if (!name) return ''
      return state.rr !== null ? `${name} · ${state.rr} RR` : name
    }
    default:
      return ''
  }
}


function applyOverrides(
  p: BuiltPresence,
  state: GameState,
  settings: Settings,
  content: Content,
  language: Language
): void {
  if (settings.largeImage !== 'auto') {
    const img = imageFor(settings.largeImage, state, content)
    if (img) p.largeImage = img
  }
  if (settings.largeText !== 'auto') {
    const txt = textFor(settings.largeText, state, content, language)
    if (txt) p.largeText = txt
  }
  if (settings.smallImage === 'none') {
    delete p.smallImage
    delete p.smallText
  } else if (settings.smallImage !== 'auto') {
    const img = imageFor(settings.smallImage, state, content)
    if (img) p.smallImage = img
  }
  if (settings.smallText !== 'auto') {
    const txt = textFor(settings.smallText, state, content, language)
    if (txt) p.smallText = txt
  }
}


export function buildPresence(
  state: GameState,
  settings: Settings,
  content: Content,
  language: Language,
  startTs: number
): BuiltPresence | null {
  if (state.sessionState === 'idle') return null

  let p: BuiltPresence
  if (state.sessionState === 'ingame') p = buildIngame(state, settings, content, language)
  else if (state.sessionState === 'pregame') p = buildPregame(state, settings, content, language)
  else p = buildMenu(state, settings, content, language)

  applyOverrides(p, state, settings, content, language)

  if (settings.showElapsed) p.startTimestamp = startTs
  if (settings.showParty && state.partySize > 0) {
    p.partySize = [state.partySize, state.partyMax]
  }
  if (settings.showButton) {
    const label = settings.buttonLabel?.trim() || 'Made by ❤️ yefeblgn'
    const url = settings.buttonUrl?.trim() || 'https://github.com/yefeblgn/valorantrpc'
    p.buttons = [{ label, url }]
  }

  return p
}


export function buildDisplay(
  state: GameState,
  content: Content
): DisplayInfo {
  const [mapName, mapSplash] = content.mapInfo(state.mapPath)
  return {
    statusKey: state.sessionState,
    isQueuing: isQueuing(state),
    isCustom: isCustom(state),
    modeName: content.modeName(state.queueId),
    mapName,
    mapSplash,
    agentName: content.agentName(state.agentUuid),
    agentIcon: state.agentUuid ? content.agentIcon(state.agentUuid) : null,
    tierName: content.tierName(state.competitiveTier),
    tierIcon: content.tierIcon(state.competitiveTier),
    cardWide: content.cardWide(state.cardId),
    cardSquare: content.cardSquare(state.cardId)
  }
}
